contrib/signet/miner

```diff
@@ -406,16 +408,33 @@ def do_generate(args):
 
         # mine block
         logging.debug("Mining block delta=%s start=%s mine=%s", seconds_to_hms(mine_time-bestheader["time"]), mine_time, is_mine)
-        mined_blocks += 1
         psbt = generate_psbt(tmpl, reward_spk, blocktime=mine_time)
         input_stream = os.linesep.join([psbt, "true", "ALL"]).encode('utf8')
         psbt_signed = json.loads(args.bcli("-stdin", "walletprocesspsbt", input=input_stream))
+
         if not psbt_signed.get("complete",False):
             logging.debug("Generated PSBT: %s" % (psbt,))
             sys.stderr.write("PSBT signing failed\n")
             return 1
         block, signet_solution = do_decode_psbt(psbt_signed["psbt"])
-        block = finish_block(block, signet_solution, args.grind_cmd)
+
+        # finish_block() will call bitcoin-util to grind PoW. If the nonce field search space is exhausted,
+        # the subprocess will fail and thus the miner. In case of failure, increment the retries var,
+        # use it to create a new block header and try again.
+        try:
+            block = finish_block(block, signet_solution, args.grind_cmd)
+        except subprocess.CalledProcessError as e:
+            logging.debug('subprocess exception: {}'.format(e))
+            logging.debug('finish_block failed, retrying')
+            retries += 1
```

Comment:
The `subprocess.CalledProcessError` could contain the stderr&out and the return code -- given there are a couple of other cases where `bitcoin-util grind` can fail, it might be good to check we actually hit an exhaustion rather than the block header decode issue to prevent many retries when something else is wrong (even if in a rare/"should never happen" case)?

```suggestion
            if e.returncode == 1 and "Could not satisfy difficulty target" in e.stderr.decode("utf-8"):
                retries += 1
            else:
                raise
```

(Maybe don't need/want to check the return code)

You'd also need to change `stdout=subprocess.PIPE` to `capture_output=True` for these to be set on the exception `e`:

```diff
diff --git a/contrib/signet/miner b/contrib/signet/miner
index 31035a5525..e50301639e 100755
--- a/contrib/signet/miner
+++ b/contrib/signet/miner
@@ -105,7 +105,7 @@ def finish_block(block, signet_solution, grind_cmd):
         headhex = CBlockHeader.serialize(block).hex()
         logging.debug('grinding headhex: {}'.format(headhex))
         cmd = grind_cmd.split(" ") + [headhex]
-        newheadhex = subprocess.run(cmd, stdout=subprocess.PIPE, input=b"", check=True).stdout.strip()
+        newheadhex = subprocess.run(cmd, capture_output=True, input=b"", check=True).stdout.strip()
         newhead = from_hex(CBlockHeader(), newheadhex.decode('utf8'))
         block.nNonce = newhead.nNonce
         block.rehash()
```

(untested with this script but I've checked a toy example so fairly sure this would work)

Reply:
Thanks, that did the job indeed.